# PyCity Schools Analysis

* As a whole, schools with higher budgets, did not yield better test results. By contrast, schools with higher spending per student actually (\$645-675) underperformed compared to schools with smaller budgets (<\$585 per student).

* As a whole, smaller and medium sized schools dramatically out-performed large sized schools on passing math performances (89-91% passing vs 67%).

* As a whole, charter schools out-performed the public district schools across all metrics. However, more analysis will be required to glean if the effect is due to school practices or the fact that charter schools tend to serve smaller student populations per school. 
---

### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies and Setup
import pandas as pd
import numpy as np

# File to Load (Remember to Change These)
school_data_to_load = "Resources/schools_complete.csv"
student_data_to_load = "Resources/students_complete.csv"

# Read School and Student Data File and store into Pandas Data Frames
school_data = pd.read_csv(school_data_to_load)
student_data = pd.read_csv(student_data_to_load)

# Combine the data into a single dataset
school_data_complete = pd.merge(student_data, school_data, how="left", on=["school_name", "school_name"])
school_data_complete.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  School ID      type  size   budget  
0             66          79          0  District  2917  1910635  
1             94          61          0  District  2917  1910635  
2             90          60          0  District  2917  1910635  
3             67          58          0  District  2917  1910635  
4             97          84          0  District  2917  1910635

## District Summary

* Calculate the total number of schools

* Calculate the total number of students

* Calculate the total budget

* Calculate the average math score 

* Calculate the average reading score

* Calculate the overall passing rate (overall average score), i.e. (avg. math score + avg. reading score)/2

* Calculate the percentage of students with a passing math score (70 or greater)

* Calculate the percentage of students with a passing reading score (70 or greater)

* Create a dataframe to hold the above results

* Optional: give the displayed data cleaner formatting

In [2]:
# Calculate total number of schools 
school_total = school_data["school_name"].count()

# Calculate total number of students
student_total = student_data["student_name"].count()

# Calculate the total budget
budget_total = school_data["budget"].sum()

# Calculate the average math and reading scores
math_average = student_data["math_score"].mean()
reading_average = student_data["reading_score"].mean()

# Calculate overall average score
overall_average = (math_average + reading_average) / 2

# Calculate the percentage of students with a passing math score (70 or greater)
math_pass_df = student_data.loc[(student_data["math_score"] >= 70)]
math_pass = math_pass_df["math_score"].count()

math_pass_rate = (math_pass / student_total)*100

# Calculate the percentage of students with a passing reading score
reading_pass_df = student_data.loc[(student_data["reading_score"] >= 70)]
reading_pass = reading_pass_df["reading_score"].count()

reading_pass_rate = (reading_pass / student_total)*100

# Calculate overall passing rate (average math and reading passing rates)
overall_pass_rate = (reading_pass_rate + math_pass_rate) / 2 

# Create a dataframe to hold the results
district_summary = pd.DataFrame(
    {"Total Schools": [school_total],
     "Total Students": [student_total],
     "Total Budget": [budget_total],
     "Average Math Score": [math_average],
     "Average Reading Score": [reading_average],
     "Math Passing Rate": [math_pass_rate],
     "Reading Passing Rate": [reading_pass_rate],
     "Overall Passing Rate": [overall_pass_rate]
     }
)

# Format table
district_summary["Total Budget"] = district_summary["Total Budget"].map("${:,}".format)
district_summary["Average Math Score"] = district_summary["Average Math Score"].map("{:.2f}".format)
district_summary["Average Reading Score"] = district_summary["Average Reading Score"].map("{:.2f}".format)
district_summary["Math Passing Rate"] = district_summary["Math Passing Rate"].map("{:.2f}%".format)
district_summary["Reading Passing Rate"] = district_summary["Reading Passing Rate"].map("{:.2f}%".format)
district_summary["Overall Passing Rate"] = district_summary["Overall Passing Rate"].map("{:.2f}%".format)

district_summary

Total Schools  Total Students Total Budget Average Math Score  \
0             15           39170  $24,649,428              78.99   

  Average Reading Score Math Passing Rate Reading Passing Rate  \
0                 81.88            74.98%               85.81%   

  Overall Passing Rate  
0               80.39%

## School Summary

* Create an overview table that summarizes key metrics about each school, including:
  * School Name
  * School Type
  * Total Students
  * Total School Budget
  * Per Student Budget
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * Overall Passing Rate (Average of the above two)
  
* Create a dataframe to hold the above results

In [3]:
# Group by school name
school_group = school_data_complete.groupby("school_name")

# Find school name 
school_name = school_group["school_name"].first() 

# Find school type 
school_type = school_group["type"].first()

# Find number of students per school
school_students = school_group["student_name"].count()

# Find school budget 
school_budget = school_group["budget"].first()

# Find school budget per student
school_student_budget = school_budget / school_students

# Find average math score per school
school_math_avg = school_group["math_score"].mean()

# Find average reading score per school
school_reading_avg = school_group["reading_score"].mean()

# Calculate the percentage of students with a passing math score (70 or greater)
school_math_pass_count = school_data_complete[school_data_complete["math_score"] >= 70].groupby("school_name")["math_score"].count() 
school_math_pass_rate = school_math_pass_count / school_students * 100

# Calculate the percentage of students with a passing reading score (70 or greater)
school_reading_pass_count = school_data_complete[school_data_complete["reading_score"] >= 70].groupby("school_name")["reading_score"].count() 
school_reading_pass_rate = school_reading_pass_count / school_students * 100

# Calculate the overall passing rate (average of reading and math scores)
school_overall_pass_rate = (school_math_pass_rate + school_reading_pass_rate) / 2

# Create a dataframe to hold the results
school_summary = pd.DataFrame(
    {"School Type": school_type,
     "Total Students": school_students,
     "Total School Budget": school_budget,
     "Per Student Budget": school_student_budget,
     "Average Math Score": school_math_avg,
     "Average Reading Score": school_reading_avg,
     "Math Passing Rate": school_math_pass_rate,
     "Reading Passing Rate": school_reading_pass_rate,
     "Overall Passing Rate": school_overall_pass_rate
     }
)

# Format table
school_summary["Total Students"] = school_summary["Total Students"].map("{:,}".format)
school_summary["Total School Budget"] = school_summary["Total School Budget"].map("${:,.0f}".format)
school_summary["Per Student Budget"] = school_summary["Per Student Budget"].map("${:,.0f}".format)
school_summary["Average Math Score"] = school_summary["Average Math Score"].map("{:.2f}".format)
school_summary["Average Reading Score"] = school_summary["Average Reading Score"].map("{:.2f}".format)
school_summary["Math Passing Rate"] = school_summary["Math Passing Rate"].map("{:.2f}%".format)
school_summary["Reading Passing Rate"] = school_summary["Reading Passing Rate"].map("{:.2f}%".format)
school_summary["Overall Passing Rate"] = school_summary["Overall Passing Rate"].map("{:.2f}%".format)

school_summary


School Type Total Students Total School Budget  \
school_name                                                            
Bailey High School       District          4,976          $3,124,928   
Cabrera High School       Charter          1,858          $1,081,356   
Figueroa High School     District          2,949          $1,884,411   
Ford High School         District          2,739          $1,763,916   
Griffin High School       Charter          1,468            $917,500   
Hernandez High School    District          4,635          $3,022,020   
Holden High School        Charter            427            $248,087   
Huang High School        District          2,917          $1,910,635   
Johnson High School      District          4,761          $3,094,650   
Pena High School          Charter            962            $585,858   
Rodriguez High School    District          3,999          $2,547,363   
Shelton High School       Charter          1,761          $1,056,600   
Thomas High School        Charter          1,635          $1,043,130   
Wilson High School        Charter          2,283          $1,319,574   
Wright High School        Charter          1,800          $1,049,400   

                      Per Student Budget Average Math Score  \
school_name                                                   
Bailey High School                  $628              77.05   
Cabrera High School                 $582              83.06   
Figueroa High School                $639              76.71   
Ford High School                    $644              77.10   
Griffin High School                 $625              83.35   
Hernandez High School               $652              77.29   
Holden High School                  $581              83.80   
Huang High School                   $655              76.63   
Johnson High School                 $650              77.07   
Pena High School                    $609              83.84   
Rodriguez High School               $637              76.84   
Shelton High School                 $600              83.36   
Thomas High School                  $638              83.42   
Wilson High School                  $578              83.27   
Wright High School                  $583              83.68   

                      Average Reading Score Math Passing Rate  \
school_name                                                     
Bailey High School                    81.03            66.68%   
Cabrera High School                   83.98            94.13%   
Figueroa High School                  81.16            65.99%   
Ford High School                      80.75            68.31%   
Griffin High School                   83.82            93.39%   
Hernandez High School                 80.93            66.75%   
Holden High School                    83.81            92.51%   
Huang High School                     81.18            65.68%   
Johnson High School                   80.97            66.06%   
Pena High School                      84.04            94.59%   
Rodriguez High School                 80.74            66.37%   
Shelton High School                   83.73            93.87%   
Thomas High School                    83.85            93.27%   
Wilson High School                    83.99            93.87%   
Wright High School                    83.95            93.33%   

                      Reading Passing Rate Overall Passing Rate  
school_name                                                      
Bailey High School                  81.93%               74.31%  
Cabrera High School                 97.04%               95.59%  
Figueroa High School                80.74%               73.36%  
Ford High School                    79.30%               73.80%  
Griffin High School                 97.14%               95.27%  
Hernandez High School               80.86%               73.81%  
Holden High School                  96.25%               94.38%  
Huang High School        

## Top Performing Schools (By Passing Rate)

* Sort and display the top five schools in overall passing rate

In [4]:
# Top Performing Schools By Passing Rate
best_schools = school_summary.sort_values("Overall Passing Rate", ascending=False)
best_schools.head(5)

School Type Total Students Total School Budget  \
school_name                                                          
Cabrera High School     Charter          1,858          $1,081,356   
Thomas High School      Charter          1,635          $1,043,130   
Griffin High School     Charter          1,468            $917,500   
Pena High School        Charter            962            $585,858   
Wilson High School      Charter          2,283          $1,319,574   

                    Per Student Budget Average Math Score  \
school_name                                                 
Cabrera High School               $582              83.06   
Thomas High School                $638              83.42   
Griffin High School               $625              83.35   
Pena High School                  $609              83.84   
Wilson High School                $578              83.27   

                    Average Reading Score Math Passing Rate  \
school_name                                                   
Cabrera High School                 83.98            94.13%   
Thomas High School                  83.85            93.27%   
Griffin High School                 83.82            93.39%   
Pena High School                    84.04            94.59%   
Wilson High School                  83.99            93.87%   

                    Reading Passing Rate Overall Passing Rate  
school_name                                                    
Cabrera High School               97.04%               95.59%  
Thomas High School                97.31%               95.29%  
Griffin High School               97.14%               95.27%  
Pena High School                  95.95%               95.27%  
Wilson High School                96.54%               95.20%

## Bottom Performing Schools (By Passing Rate)

* Sort and display the five worst-performing schools

In [5]:
# Bottom Performing Schools By Passing Rate
worst_schools = school_summary.sort_values("Overall Passing Rate")
worst_schools.head(5)

School Type Total Students Total School Budget  \
school_name                                                            
Rodriguez High School    District          3,999          $2,547,363   
Figueroa High School     District          2,949          $1,884,411   
Huang High School        District          2,917          $1,910,635   
Johnson High School      District          4,761          $3,094,650   
Ford High School         District          2,739          $1,763,916   

                      Per Student Budget Average Math Score  \
school_name                                                   
Rodriguez High School               $637              76.84   
Figueroa High School                $639              76.71   
Huang High School                   $655              76.63   
Johnson High School                 $650              77.07   
Ford High School                    $644              77.10   

                      Average Reading Score Math Passing Rate  \
school_name                                                     
Rodriguez High School                 80.74            66.37%   
Figueroa High School                  81.16            65.99%   
Huang High School                     81.18            65.68%   
Johnson High School                   80.97            66.06%   
Ford High School                      80.75            68.31%   

                      Reading Passing Rate Overall Passing Rate  
school_name                                                      
Rodriguez High School               80.22%               73.29%  
Figueroa High School                80.74%               73.36%  
Huang High School                   81.32%               73.50%  
Johnson High School                 81.22%               73.64%  
Ford High School                    79.30%               73.80%

## Math Scores by Grade

* Create a table that lists the average Reading Score for students of each grade level (9th, 10th, 11th, 12th) at each school.

  * Create a pandas series for each grade. Hint: use a conditional statement.
  
  * Group each series by school
  
  * Combine the series into a dataframe
  
  * Optional: give the displayed data cleaner formatting

In [6]:
# Create a pandas series for each grade.
ninth_grade = school_data_complete.loc[school_data_complete["grade"] == "9th"] 
tenth_grade = school_data_complete.loc[school_data_complete["grade"] == "10th"]
eleventh_grade = school_data_complete.loc[school_data_complete["grade"] == "11th"] 
twelfth_grade = school_data_complete.loc[school_data_complete["grade"] == "12th"]

# Calculate average math scores by school 
school_ninth_math_avg = round(ninth_grade.groupby("school_name")["math_score"].mean(), 2)
school_tenth_math_avg = round(tenth_grade.groupby("school_name")["math_score"].mean(), 2)
school_eleventh_math_avg = round(eleventh_grade.groupby("school_name")["math_score"].mean(), 2)
school_twelfth_math_avg = round(twelfth_grade.groupby("school_name")["math_score"].mean(), 2)

# Create a dataframe to hold the results
math_by_grade = pd.DataFrame(
    {"9th Grade Math Score Average": school_ninth_math_avg,
     "10th Grade Math Score Average": school_tenth_math_avg,
     "11th Grade Math Score Average": school_eleventh_math_avg,
     "12th Grade Math Score Average": school_twelfth_math_avg
     }
)
math_by_grade


9th Grade Math Score Average  \
school_name                                           
Bailey High School                            77.08   
Cabrera High School                           83.09   
Figueroa High School                          76.40   
Ford High School                              77.36   
Griffin High School                           82.04   
Hernandez High School                         77.44   
Holden High School                            83.79   
Huang High School                             77.03   
Johnson High School                           77.19   
Pena High School                              83.63   
Rodriguez High School                         76.86   
Shelton High School                           83.42   
Thomas High School                            83.59   
Wilson High School                            83.09   
Wright High School                            83.26   

                       10th Grade Math Score Average  \
school_name                                            
Bailey High School                             77.00   
Cabrera High School                            83.15   
Figueroa High School                           76.54   
Ford High School                               77.67   
Griffin High School                            84.23   
Hernandez High School                          77.34   
Holden High School                             83.43   
Huang High School                              75.91   
Johnson High School                            76.69   
Pena High School                               83.37   
Rodriguez High School                          76.61   
Shelton High School                            82.92   
Thomas High School                             83.09   
Wilson High School                             83.72   
Wright High School                             84.01   

                       11th Grade Math Score Average  \
school_name                                            
Bailey High School                             77.52   
Cabrera High School                            82.77   
Figueroa High School                           76.88   
Ford High School                               76.92   
Griffin High School                            83.84   
Hernandez High School                          77.14   
Holden High School                             85.00   
Huang High School                              76.45   
Johnson High School                            77.49   
Pena High School                               84.33   
Rodriguez High School                          76.40   
Shelton High School                            83.38   
Thomas High School                             83.50   
Wilson High School                             83.20   
Wright High School                             83.84   

                       12th Grade Math Score Average  
school_name                                           
Bailey High School                             76.49  
Cabrera High School                            83.28  
Figueroa High School                           77.15  
Ford High School                               76.18  
Griffin High School                            83.36  
Hernandez High School                          77.19  
Holden High School                             82.86  
Huang High School                              77.23  
Johnson High School                            76.86  
Pena High School                               84.12  
Rodriguez High School                          77.69  
Shelton High School                            83.78  
Thomas High School                             83.50  
Wilson High School                             83.04  
Wright High School                             83.64

## Reading Score by Grade 

* Perform the same operations as above for reading scores

In [7]:
# Series for each grade created in previous step
# Calculate average reading scores by school
school_ninth_reading_avg = round(ninth_grade.groupby("school_name")["reading_score"].mean(), 2)
school_tenth_reading_avg = round(tenth_grade.groupby("school_name")["reading_score"].mean(), 2)
school_eleventh_reading_avg = round(eleventh_grade.groupby("school_name")["reading_score"].mean(), 2)
school_twelfth_reading_avg = round(twelfth_grade.groupby("school_name")["reading_score"].mean(), 2)

# Create a dataframe to hold the results
reading_by_grade = pd.DataFrame(
    {"9th Grade Reading Score Average": school_ninth_reading_avg,
     "10th Grade Reading Score Average": school_tenth_reading_avg,
     "11th Grade Reading Score Average": school_eleventh_reading_avg,
     "12th Grade Reading Score Average": school_twelfth_reading_avg
     }
)
reading_by_grade

9th Grade Reading Score Average  \
school_name                                              
Bailey High School                               81.30   
Cabrera High School                              83.68   
Figueroa High School                             81.20   
Ford High School                                 80.63   
Griffin High School                              83.37   
Hernandez High School                            80.87   
Holden High School                               83.68   
Huang High School                                81.29   
Johnson High School                              81.26   
Pena High School                                 83.81   
Rodriguez High School                            80.99   
Shelton High School                              84.12   
Thomas High School                               83.73   
Wilson High School                               83.94   
Wright High School                               83.83   

                       10th Grade Reading Score Average  \
school_name                                               
Bailey High School                                80.91   
Cabrera High School                               84.25   
Figueroa High School                              81.41   
Ford High School                                  81.26   
Griffin High School                               83.71   
Hernandez High School                             80.66   
Holden High School                                83.32   
Huang High School                                 81.51   
Johnson High School                               80.77   
Pena High School                                  83.61   
Rodriguez High School                             80.63   
Shelton High School                               83.44   
Thomas High School                                84.25   
Wilson High School                                84.02   
Wright High School                                83.81   

                       11th Grade Reading Score Average  \
school_name                                               
Bailey High School                                80.95   
Cabrera High School                               83.79   
Figueroa High School                              80.64   
Ford High School                                  80.40   
Griffin High School                               84.29   
Hernandez High School                             81.40   
Holden High School                                83.82   
Huang High School                                 81.42   
Johnson High School                               80.62   
Pena High School                                  84.34   
Rodriguez High School                             80.86   
Shelton High School                               84.37   
Thomas High School                                83.59   
Wilson High School                                83.76   
Wright High School                                84.16   

                       12th Grade Reading Score Average  
school_name                                              
Bailey High School                                80.91  
Cabrera High School                               84.29  
Figueroa High School                              81.38  
Ford High School                                  80.66  
Griffin High School                               84.01  
Hernandez High School                             80.86  
Holden High School                                84.70  
Huang High School                                 80.31  
Johnson High School                               81.23  
Pena High School                                  84.59  
Rodriguez High School                             80.38  
Shelton High School                               82.78  
Thomas High School                                83.83  
Wilson High School                                84.32  
Wright High School                                84.07

## Scores by School Spending

* Create a table that breaks down school performances based on average Spending Ranges (Per Student). Use 4 reasonable bins to group school spending. Include in the table each of the following:
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * Overall Passing Rate (Average of the above two)

In [8]:
# Sample bins. Feel free to create your own bins.
spending_bins = [0, 585, 615, 645, 675]
group_names = ["<$585", "$585-615", "$615-645", "$645-675"]

# Add spending ranges to the dataframe
school_data_complete["budget_per_student"] = pd.cut(school_data_complete["budget"] / school_data_complete["size"], spending_bins, labels=group_names)
school_data_complete.head()



Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  School ID      type  size   budget  \
0             66          79          0  District  2917  1910635   
1             94          61          0  District  2917  1910635   
2             90          60          0  District  2917  1910635   
3             67          58          0  District  2917  1910635   
4             97          84          0  District  2917  1910635   

  budget_per_student  
0           $645-675  
1           $645-675  
2           $645-675  
3           $645-675  
4           $645-675

In [9]:
# Group by spending bins
spending_group = school_data_complete.groupby("budget_per_student")

# Find number of students per spending group
spending_students = spending_group["student_name"].count()

# Calculate average math score
spending_math_avg = spending_group["math_score"].mean()

# Find average reading score per school
spending_reading_avg = spending_group["reading_score"].mean()

# Calculate the percentage of students with a passing math score (70 or greater)
spending_math_pass_count = school_data_complete[school_data_complete["math_score"] >= 70].groupby("budget_per_student")["math_score"].count() 
spending_math_pass_rate = spending_math_pass_count / spending_students * 100

# Calculate the percentage of students with a passing reading score (70 or greater)
spending_reading_pass_count = school_data_complete[school_data_complete["reading_score"] >= 70].groupby("budget_per_student")["reading_score"].count() 
spending_reading_pass_rate = spending_reading_pass_count / spending_students * 100

# Calculate the overall passing rate (average of reading and math scores)
spending_overall_pass_rate = (spending_math_pass_rate + spending_reading_pass_rate) / 2

# Create a dataframe to hold the results
spending_summary = pd.DataFrame(
    {"Average Math Score": spending_math_avg,
     "Average Reading Score": spending_reading_avg,
     "Math Passing Rate": spending_math_pass_rate,
     "Reading Passing Rate": spending_reading_pass_rate,
     "Overall Passing Rate": spending_overall_pass_rate
     }
)

# Format table
spending_summary["Average Math Score"] = spending_summary["Average Math Score"].map("{:.2f}".format)
spending_summary["Average Reading Score"] = spending_summary["Average Reading Score"].map("{:.2f}".format)
spending_summary["Math Passing Rate"] = spending_summary["Math Passing Rate"].map("{:.2f}%".format)
spending_summary["Reading Passing Rate"] = spending_summary["Reading Passing Rate"].map("{:.2f}%".format)
spending_summary["Overall Passing Rate"] = spending_summary["Overall Passing Rate"].map("{:.2f}%".format)

spending_summary


Average Math Score Average Reading Score Math Passing Rate  \
budget_per_student                                                              
<$585                           83.36                 83.96            93.70%   
$585-615                        83.53                 83.84            94.12%   
$615-645                        78.06                 81.43            71.40%   
$645-675                        77.05                 81.01            66.23%   

                   Reading Passing Rate Overall Passing Rate  
budget_per_student                                            
<$585                            96.69%               95.19%  
$585-615                         95.89%               95.01%  
$615-645                         83.61%               77.51%  
$645-675                         81.11%               73.67%

## Scores by School Size

* Perform the same operations as above, based on school size.

In [10]:
# Sample bins. Feel free to create your own bins.
size_bins = [0, 1000, 2000, 5000]
group_names = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]

# Add size ranges to the dataframe
school_data_complete["size_range"] = pd.cut(school_data_complete["size"], size_bins, labels=group_names)
school_data_complete.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  School ID      type  size   budget  \
0             66          79          0  District  2917  1910635   
1             94          61          0  District  2917  1910635   
2             90          60          0  District  2917  1910635   
3             67          58          0  District  2917  1910635   
4             97          84          0  District  2917  1910635   

  budget_per_student         size_range  
0           $645-675  Large (2000-5000)  
1           $645-675  Large (2000-5000)  
2           $645-675  Large (2000-5000)  
3           $645-675  Large (2000-5000)  
4           $645-675  Large (2000-5000)

In [11]:
# Group by size bins
size_group = school_data_complete.groupby("size_range")

# Find number of students per size group
size_students = size_group["student_name"].count()

# Calculate average math score
size_math_avg = size_group["math_score"].mean()

# Find average reading score per school
size_reading_avg = size_group["reading_score"].mean()

# Calculate the percentage of students with a passing math score (70 or greater)
size_math_pass_count = school_data_complete[school_data_complete["math_score"] >= 70].groupby("size_range")["math_score"].count() 
size_math_pass_rate = size_math_pass_count / size_students * 100

# Calculate the percentage of students with a passing reading score (70 or greater)
size_reading_pass_count = school_data_complete[school_data_complete["reading_score"] >= 70].groupby("size_range")["reading_score"].count() 
size_reading_pass_rate = size_reading_pass_count / size_students * 100

# Calculate the overall passing rate (average of reading and math scores)
size_overall_pass_rate = (size_math_pass_rate + size_reading_pass_rate) / 2

# Create a dataframe to hold the results
size_summary = pd.DataFrame(
    {"Average Math Score": size_math_avg,
     "Average Reading Score": size_reading_avg,
     "Math Passing Rate": size_math_pass_rate,
     "Reading Passing Rate": size_reading_pass_rate,
     "Overall Passing Rate": size_overall_pass_rate
     }
)

# Format table
size_summary["Average Math Score"] = size_summary["Average Math Score"].map("{:.2f}".format)
size_summary["Average Reading Score"] = size_summary["Average Reading Score"].map("{:.2f}".format)
size_summary["Math Passing Rate"] = size_summary["Math Passing Rate"].map("{:.2f}%".format)
size_summary["Reading Passing Rate"] = size_summary["Reading Passing Rate"].map("{:.2f}%".format)
size_summary["Overall Passing Rate"] = size_summary["Overall Passing Rate"].map("{:.2f}%".format)

size_summary

Average Math Score Average Reading Score Math Passing Rate  \
size_range                                                                      
Small (<1000)                   83.83                 83.97            93.95%   
Medium (1000-2000)              83.37                 83.87            93.62%   
Large (2000-5000)               77.48                 81.20            68.65%   

                   Reading Passing Rate Overall Passing Rate  
size_range                                                    
Small (<1000)                    96.04%               95.00%  
Medium (1000-2000)               96.77%               95.19%  
Large (2000-5000)                82.13%               75.39%

## Scores by School Type

* Perform the same operations as above, based on school type.

In [12]:
# Group by size bins
type_group = school_data_complete.groupby("type")

# Find number of students per spending group
type_students = type_group["student_name"].count()

# Calculate average math score
type_math_avg = type_group["math_score"].mean()

# Find average reading score per school
type_reading_avg = type_group["reading_score"].mean()

# Calculate the percentage of students with a passing math score (70 or greater)
type_math_pass_count = school_data_complete[school_data_complete["math_score"] >= 70].groupby("type")["math_score"].count() 
type_math_pass_rate = type_math_pass_count / type_students * 100

# Calculate the percentage of students with a passing reading score (70 or greater)
type_reading_pass_count = school_data_complete[school_data_complete["reading_score"] >= 70].groupby("type")["reading_score"].count() 
type_reading_pass_rate = type_reading_pass_count / type_students * 100

# Calculate the overall passing rate (average of reading and math scores)
type_overall_pass_rate = (type_math_pass_rate + type_reading_pass_rate) / 2

# Create a dataframe to hold the results
type_summary = pd.DataFrame(
    {"Average Math Score": type_math_avg,
     "Average Reading Score": type_reading_avg,
     "Math Passing Rate": type_math_pass_rate,
     "Reading Passing Rate": type_reading_pass_rate,
     "Overall Passing Rate": type_overall_pass_rate
     }
)

# Format table
type_summary["Average Math Score"] = type_summary["Average Math Score"].map("{:.2f}".format)
type_summary["Average Reading Score"] = type_summary["Average Reading Score"].map("{:.2f}".format)
type_summary["Math Passing Rate"] = type_summary["Math Passing Rate"].map("{:.2f}%".format)
type_summary["Reading Passing Rate"] = type_summary["Reading Passing Rate"].map("{:.2f}%".format)
type_summary["Overall Passing Rate"] = type_summary["Overall Passing Rate"].map("{:.2f}%".format)

type_summary

Average Math Score Average Reading Score Math Passing Rate  \
type                                                                  
Charter               83.41                 83.90            93.70%   
District              76.99                 80.96            66.52%   

         Reading Passing Rate Overall Passing Rate  
type                                                
Charter                96.65%               95.17%  
District               80.91%               73.71%

### Analysis

1) Large schools (with 2,000 - 5,000 students) have lower pass rates in both reading and math than small- (less than 1,000 students) and medium (between 1,000 - 2,000 students) schools, which have roughly the same pass rates in reading and math. Math passing rates are especially poor in large schools, with a roughly 69% pass rate, on average, compared to 94% for small and medium schools.

2) In this sample, charter schools have much higher overall passing rates than district public schools. Charter schools don't see a big drop off in the pass rate between reading and math, which have 97% and 94% pass rates, respectively. But district schools, which have lower pass rates in both subjects, see a large drop off in the math pass rate (67%) versus the reading pass rate (81%). 
